Add convenience functions to IndexSet

gassmoeller opened the pull request:
This PR adds a number of convenience functions to the `IndexSet` class, when one is simply interested in getting the first or the last index in the IndexSet and remove it at the same time from the list of indices.

It also allows to quickly check if the IndexSet is empty (much faster than `n_elements() == 0` because it does not need to compress the IndexSet first).

bangerth commented on source/base/index_set.cc:
```diff
@@ -341,6 +341,44 @@ IndexSet::subtract_set (const IndexSet &other)
+  if (ranges.front().begin == ranges.front().end)
+    ranges.erase(ranges.begin());
+
+  // We have to set this in any case, because nth_index_in_set is not longer
```
not -> no

bangerth commented on source/base/index_set.cc:
```diff
@@ -341,6 +341,44 @@ IndexSet::subtract_set (const IndexSet &other)
+IndexSet::size_type
+IndexSet::pop_back ()
+{
+  Assert(ranges.size() != 0,
```
couldn't you just call `is_empty() == false` here?

bangerth commented on source/base/index_set.cc:
```diff
@@ -341,6 +341,44 @@ IndexSet::subtract_set (const IndexSet &other)
+IndexSet::size_type
+IndexSet::pop_front ()
+{
+  Assert(ranges.size() != 0,
```
same here. I think that would be more obvious.

gassmoeller commented:
I addressed the comments. I wrote `is_empty` after the other functions (when I noticed `n_elements()` can be slow, if it needs to compress ranges). It indeed makes more sense to use it now.

bangerth commented:
/run-tests

bangerth commented:
OK to merge once the tester is happy.
